felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -1314,13 +1363,171 @@ func (m *bpfEndpointManager) setAcceptLocal(iface string, val bool) error {
+		}
+		if err := nl.LinkSetUp(link); err != nil {
+			log.WithError(err).Fatal("Failed to set bpfnatin up.")
+		}
+		link, err = nl.LinkByName("bpfnatout")
+		if err != nil {
+			log.WithError(err).Fatal("Miss bpfnatout after add.")
+		}
+		if err := nl.LinkSetUp(link); err != nil {
+			log.WithError(err).Fatal("Failed to set bpfnatout up.")
+		}
+	}
+
+	if err := configureInterface("bpfnatin", 4, writeProcSys); err != nil {
+		log.WithError(err).Fatal("Failed to configure bpfnatin parameters.")
+	}
+	if err := configureInterface("bpfnatout", 4, writeProcSys); err != nil {
+		log.WithError(err).Fatal("Failed to configure bpfnatout parameters.")
+	}
+
+	err = m.ensureQdisc("bpfnatin")
+	if err != nil {
+		log.WithError(err).Fatalf("Failed to set qdisc on bpfnatin.")
+	}
+
+	if err := m.dp.setRPFilter("bpfnatin", 0); err != nil {
+		log.WithError(err).Fatalf("Failed to disable RPF on bpfnatin.")
+	}
+	if err := m.dp.setRPFilter("bpfnatout", 0); err != nil {
+		log.WithError(err).Fatalf("Failed to disable RPF on bpfnatout.")
+	}
+
+	log.Info("Created bpfnatin pair.")
+
+	m.rulesMakeSpaceForNATIface(nl)
+
+	m.routeTableSNAT = routetable.New(
+		nil,
+		4,
+		false, // vxlan
+		10*time.Second,
+		nil, // deviceRouteSourceAddress
+		1,
+		true, // removeExternalRoutes
+		m.routeTableSNATIdx,
+		m.opReporter,
+	)
+
+	anyV4, _ := ip.CIDRFromString("0.0.0.0/0")
+	m.routeTableSNAT.RouteUpdate("bpfnatout", routetable.Target{
+		Type: routetable.TargetTypeUnicast,
+		CIDR: anyV4,
+	})
+
+	m.routeRuleSNAT, err = routerule.New(
+		4,
+		0, // routing priority
+		set.From(m.routeTableSNATIdx),
+		routerule.RulesMatchSrcFWMarkTable,
+		routerule.RulesMatchSrcFWMarkTable,
+		10*time.Second,
+		func() (routerule.HandleIface, error) {
+			return netlinkshim.NewRealNetlink()
+		},
+		m.opReporter,
+	)
+	if err != nil {
+		log.WithError(err).Panic("Unexpected error creating rule manager")
+	}
+	m.routeRuleSNAT.SetRule(routerule.NewRule(4, 0).
+		GoToTable(m.routeTableSNATIdx).
+		MatchFWMarkWithMask(tcdefs.MarkSeenToNatIfaceOut, tcdefs.MarkSeenToNatIfaceOut),
+	)
+}
+
+func (m *bpfEndpointManager) rulesMakeSpaceForNATIface(nl netlinkshim.Interface) {
+	rules, err := nl.RuleList(unix.AF_INET)
+	if err != nil {
+		log.WithError(err).Fatalf("Failed to list routing rules.")
+	}
+	log.WithField("rules", rules).Debugf("Existing routing rules")
+
+	sort.SliceStable(rules, func(i, j int) bool { return rules[i].Priority < rules[j].Priority })
+
+	if len(rules) == 0 || rules[0].Priority > 1 {
+		return
+	}
+
+	i := 0
+
+	// Skip over the (one?) default rules
+	for ; rules[i].Priority < 0; i++ {
+	}
+
+	// Is there already a gap?
+	for ; rules[i].Priority < 1; i++ {
+	}
+
+	// If there are rules at the start, find a gap of 2 (as the -1 needs to become 1 so that we can enter 0)
+	if i > 0 {
+		for ; i < len(rules); i++ {
+			if rules[i].Priority-rules[i-1].Priority > 2 {
+				break
+			}
+		}
+	}
+
+	// Shift the rules
+	for i--; i >= 0; i-- {
```

Comment:
This could shift other Calico rules around, depending on configuration; then we'd fight with the logic that owns those rules.  We'll need to address that.